### PR TITLE
Expose programmatic suppressions from SuppressionInfo

### LINKED ISSUE #69000: Make ProgrammaticSuppressionInfo public
## Background and Motivation

We added support for `DiagnosticSuppressor` API some time back. This work added a new internal type `ProgrammaticSuppressionInfo` and an internal property `Diagnostic.ProgrammaticSuppressionInfo` to the compiler layer. This issue proposes making these two APIs public, ~~without any changes or additions to the implementation~~ with one minor change to the existing internal API shape: change the type of [ProgrammaticSuppressionInfo.Suppressions](https://github.com/dotnet/roslyn/blob/eee6243a578fa52f88bc34ccbfa33356797fef94/src/Compilers/Core/Portable/Diagnostic/ProgrammaticSuppressionInfo.cs#L15) property from `ImmutableHashSet<(string Id, LocalizableString Justification)>` to `ImmutableArray<Suppression>`. See existing definition of public type `Suppression` [here](https://github.com/dotnet/roslyn/blob/main/src/Compilers/Core/Portable/DiagnosticAnalyzer/Suppression.cs)

**Motivation:**
1. Testing support: From @sharwell:
      > We need this public API for `DiagnosticSuppressor` unit testing support in the Microsoft.CodeAnalysis.Testing SDK. We tried two other approaches using public APIs but they caused problems for various edge cases. The library now uses the current property via reflection, which is fragile.

2. Tooling support: Tools built on top of Microsoft.CodeAnalysis that require to read/operate on the programmatic suppressions from `DiagnosticSuppressors` would need this public API. 

### Existing internal type [ProgrammaticSuppressionInfo](https://github.com/dotnet/roslyn/blob/main/src/Compilers/Core/Portable/Diagnostic/ProgrammaticSuppressionInfo.cs) to be made public:

```diff
namespace Microsoft.CodeAnalysis.Diagnostics
{
+     public sealed class ProgrammaticSuppressionInfo : IEquatable<ProgrammaticSuppressionInfo?>
+     {
+        // NOTE: The below property is currently implemented in Roslyn with type `ImmutableHashSet<(string Id, LocalizableString Justification)>`, which might not be best suited for a public API.
+        //       Also note that this is proposed as an array and not a single Suppression instance because multiple diagnostic suppressors can suppress the same diagnostic. For most real world cases, the below array will have a single element.
+        public ImmutableArray<Suppression> Suppressions { get; }

+        public bool Equals(ProgrammaticSuppressionInfo? other);
+        public overrides bool Equals(object? other);
+        public overrides int GetHashCode();
+     }
}
```

### Existing internal property [Diagnostic.ProgrammaticSuppressionInfo](https://github.com/dotnet/roslyn/blob/86156faa88860dac9b17223623743aea24814291/src/Compilers/Core/Portable/Diagnostic/Diagnostic.cs#L474) to be made public:

```diff
namespace Microsoft.CodeAnalysis
{
      public abstract class Diagnostic
      {
+        public virtual ProgrammaticSuppressionInfo? ProgrammaticSuppressionInfo { get; };
      }
}
```

**NOTE:** `Diagnostic` type has internal abstract methods, and hence cannot be sub-typed outside the compiler layer: https://github.com/dotnet/roslyn/blob/86156faa88860dac9b17223623743aea24814291/src/Compilers/Core/Portable/Diagnostic/Diagnostic.cs#L448-L461
```

### PR DESCRIPTION
**Pull request by mavasani**
Closes #69000

Implements the proposal in https://github.com/dotnet/roslyn/issues/69000#issuecomment-1664646444

### CODE REVIEW

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs [hunk 1]
```diff
@@ -1151,24 +1151,27 @@ ImmutableArray<Diagnostic> getSuppressableDiagnostics(DiagnosticSuppressor suppr
...
+                    if (!programmaticSuppressionsBuilder.TryGetValue(programmaticSuppression.SuppressedDiagnostic, out var array))
                     {
-                        set = ImmutableHashSet.CreateBuilder<(string, LocalizableString)>();
-                        programmaticSuppressionsBuilder.Add(programmaticSuppression.SuppressedDiagnostic, set);
+                        array = ArrayBuilder<Suppression>.GetInstance();
+                        programmaticSuppressionsBuilder.Add(programmaticSuppression.SuppressedDiagnostic, array);
                     }
 
-                    set.Add((programmaticSuppression.Descriptor.Id, programmaticSuppression.Descriptor.Justification));
+                    if (!array.Contains(programmaticSuppression))
```
Comment (mavasani) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs:1164]: This array should almost always have a single element as it is unlikely to have multiple suppressors suppressing the same diagnostic, so the Contains call shouldn't add to any cost here.

**Review by 333fred — APPROVED** [review]

**Comment by mavasani** [comment]
@dotnet/roslyn-compiler for second review

**Comment by jaredpar** [comment]
@cston PTAL

**Review by cston — APPROVED** [review]

**Comment by mavasani** [comment]
Thanks!

**Pull request merged**